src/identity_committer.rs

```diff
@@ -128,61 +120,33 @@ impl IdentityCommitter {
     async fn commit_identity(
         database: &Database,
         contracts: &Contracts,
-        tree_state: &TimedReadProgressLock<TreeState>,
+        index: &AtomicI64,
         group_id: usize,
         commitment: Hash,
     ) -> AnyhowResult<()> {
-        // Get a progress lock on the tree for the duration of this operation. Holding a
-        // progress lock ensures no other thread calls tries to insert an identity into
-        // the contract, as that is an order dependent operation.
-        let tree = tree_state.progress().await.map_err(|e| {
-            error!(?e, "Failed to obtain tree lock in commit_identity.");
-            e
-        })?;
-
         // Send Semaphore transaction
         let receipt = contracts.insert_identity(commitment).await.map_err(|e| {
             error!(?e, "Failed to insert identity to contract.");
             e
         })?;
 
-        let mut tree = tree.upgrade_to_write().await.map_err(|e| {
-            error!(?e, "Failed to obtain tree lock in insert_identity.");
-            e
-        })?;
-
-        // Update  merkle tree
-        let identity_index = tree.next_leaf;
-        tree.merkle_tree.set(identity_index, commitment);
-        tree.next_leaf += 1;
-
-        // Downgrade write lock to progress lock – tree is now up to date, but still
-        // need to update the database.
-        let tree = tree.downgrade_to_progress();
-
         let block = receipt
             .block_number
             .expect("Transaction is mined, block number must be present.");
-        info!(
-            "Identity inserted in block {} at index {}.",
-            block, identity_index
-        );
+
+        info!("Identity submitted in block {}.", block);
         database
-            .mark_identity_inserted(group_id, &commitment, block.as_usize(), identity_index)
+            .mark_identity_inserted(
+                group_id,
+                &commitment,
+                block.as_usize(),
+                index.fetch_add(1, Ordering::AcqRel),
```

Comment:
This is mighty strange, a counter that is not at all restart-resilient? Why do we care at all?

Reply:
What I'm saying is: remove it from the DB, what uses this?

Reply:
You are right, this was completely wrong. I removed the index column and simply depend on created_at order instead. This should be simpler and restart-resilient.